### PR TITLE
Add support for the 'findexpr' option

### PR DESCRIPTION
**Pull request by yegappan**
Add support for 'findexpr' which is evaluated when searching for file names using the ":find" command.

### CODE REVIEW

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -3552,6 +3552,31 @@ A jump table for the options with a short description can be found at |Q_op|.
...
 	  lastline	NonText			|hl-NonText|
 
+						*'findexpr'* *'fexpr'*
+'findexpr' 'fexpr'	string	(default "")
+			global or local to buffer |global-local|
+	Expression that is evaluated to obtain the filename for the |:find|
+	command.  When this option is empty, the internal |file-searching|
+	mechanism is used.
+	If a match is found, the expression should return a String containing
+	one or more file names separated by a newline character.  While
```
Comment (benknoble) [inline runtime/doc/options.txt]: This is probably not robust enough. The expr should return a list of filenames: that way filenames with newlines are handled without issue, and systemlist() becomes easier to use.

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -3552,6 +3552,31 @@ A jump table for the options with a short description can be found at |Q_op|.
 	  eob		EndOfBuffer		|hl-EndOfBuffer|
 	  lastline	NonText			|hl-NonText|
 
+						*'findexpr'* *'fexpr'*
```
Comment (benknoble) [inline runtime/doc/options.txt:3555]: Might be worth mentioning that custom exprs can still support `path` if they choose?

File: src/ex_docmd.c [hunk 3]
```diff
@@ -6923,6 +6923,111 @@ ex_wrongmodifier(exarg_T *eap)
...
 }
 
+#ifdef FEAT_EVAL
+/*
+ * Skip to the next newline character.
+ */
+    static char_u *
+skiptonewline(char_u *p)
+{
+    while (*p && *p != NL)
```
Comment (benknoble) [inline src/ex_docmd.c]: Can this not use something like `memchr`? But see earlier comment about not parsing newline-delimited filenames

**Comment by chrisbra** [comment]
yes, I think returning a list of files would make it more robust.

File: src/testdir/test_modeline.vim [hunk 4]
```diff
@@ -258,6 +258,7 @@ func Test_modeline_fails_always()
   call s:modeline_fails('t_ce', 't_ce=Something()', 'E520:')
   call s:modeline_fails('t_sr', 't_sr=Something()', 'E520:')
   call s:modeline_fails('t_8b', 't_8b=Something()', 'E520:')
+  call s:modeline_fails('findexpr', 'findexpr=Something()', 'E520:')
```
Comment (zeertzjq) [inline src/testdir/test_modeline.vim]: This should come before the `" only check a few terminal options` comment, and probably between `formatprg` and `exrc`

File: runtime/doc/options.txt (hunk 1 above)
Comment (yegappan) [reply runtime/doc/options.txt]: The PR is updated to support this.

File: src/ex_docmd.c (hunk 3 above)
Comment (yegappan) [reply src/ex_docmd.c]: This function is removed in the updated PR.

File: runtime/doc/quickref.txt [hunk 5]
```diff
@@ -707,6 +707,7 @@ Short explanation of each option:		*option-list*
 'fileignorecase'  'fic'     ignore case when using file names
 'filetype'	  'ft'	    type of file, used for autocommands
 'fillchars'	  'fcs'     characters to use for displaying special items
+'findexpr'	  'fexpr'   expression to evaluate for|:find|
```
Comment (zeertzjq) [inline runtime/doc/quickref.txt:710]:
```suggestion
'findexpr'	  'fexpr'   expression to evaluate for |:find|
```

File: runtime/doc/options.txt (hunk 1 above)
Comment (benknoble) [reply runtime/doc/options.txt]: Thanks; I don’t have permissions here to resolve my own review comment. Please close this and the memchr comment if possible.

**Comment by chrisbra** [comment]
Thanks. I assume making `expand_path_option` non-static was an old try? I'll leave this part out when merging.

**Comment by justinmk** [comment]
Why not 'findfunc' (which accepts a funcref/lambda)? Why add new "expr" options?

**Comment by chrisbra** [comment]
honestly I did not think about 'func' options.

**Comment by yegappan** [comment]
> Why not 'findfunc' (which accepts a funcref/lambda)? Why add new "expr" options?

There is no particular reason.  In fact, yesterday I was trying to add a test to use a lambda
for `findexpr`.  I realized we should have added `funcfunc` instead of `findexpr`.

**Comment by chrisbra** [comment]
yeah, it's still quite new, so  I think we can make a `findfunc` option out of it, if we all agree.

**Pull request closed without merging**